tactics/proof_global.ml

```diff
@@ -164,31 +168,22 @@ let return_proof { proof } =
      equations and so far there is no code in the CI that will
      actually call those and do a side-effect, TTBOMK *)
   (* EJGA: likely the right solution is to attach side effects to the first constant only? *)
-  let proofs = List.map (fun (c, _) -> (proof_opt c, eff)) initial_goals in
+  let proofs = List.map (fun (body, typ) -> (to_constr_body body, eff), to_constr_typ typ) initial_goals in
   proofs, Evd.evar_universe_context evd
 
 let close_proof ~opaque ~keep_body_ucst_separate ps =
-  let elist, uctx = return_proof ps in
+  let elist, uctx = prepare_proof ~unsafe_typ:keep_body_ucst_separate ps in
```

Comment:
How confident are we to change from `not poly && (keep_body_ucst_separate || not (Safe_typing.empty_private_constants = eff.Evd.seff_private))` to just `keep_body_ucst_separate`?

Reply:
I mean here we change behaviour (when `poly && keep_body_ucst_separate` we now use unafe_typ), is that really OK?

Reply:
Ah I see what you mean; in theory there is an invariant on the STM that in theory doesn't allow to set `poly=true` and `keep_body_separate = true`.

Actually the original PR was enforcing this invariant by splitting the function for the vio case, however I found other problems so that's why I didn't care so much; let's strengthen the condition for now.